### ISSUE TITLE
Not enough information to get shape

### ISSUE THREAD

**Issue by nnguyen24** [apache/incubator-mxnet#6629]
Hello,

I am implementing a neural network in MXNetR. I attempted to customize my loss function to compute the correlation between my output vector and the targeting vector. Below is my code:

Below is my code:
```
    # Generate testing data
    train.x = matrix(data = rexp(200, rate = 10), nrow = 120, ncol = 6380)
    test.x = matrix(data = rexp(200, rate = 10), nrow = 60, ncol = 6380)
    train.y = matrix(data = rexp(200, rate = 10), nrow = 120, ncol = 319)
    test.y = matrix(data = rexp(200, rate = 10), nrow = 60, ncol = 319)
   
    # Reshape testing data
    train.array <-train.x
    dim(train.array) <-c(20,319,1,ncol(train.x))
    test.array<-test.x
    dim(test.array) <-c (20,319,1,ncol(test.x))

    # Define the input data
    data <- mx.symbol.Variable("data")
    
    # Define the first fully connected layer
    fc1 <- mx.symbol.FullyConnected(data, num_hidden = 100)
    act.fun <- mx.symbol.Activation(fc1, act_type = "relu") # create a hidden layer with Rectified Linear Unit as its activation function.
    output <<- mx.symbol.FullyConnected(act.fun, num_hidden = 319)
    
    # Customize loss function
    label <- mx.symbol.Variable("label")
   
    output_mean <- mx.symbol.mean(output)
    label_mean <- mx.symbol.mean(label)

    output_delta <-mx.symbol.broadcast_sub(output, output_mean)
    label_delta <- mx.symbol.broadcast_sub(label, label_mean)

    output_sqr <-mx.symbol.square(output_delta)
    label_sqr <- mx.symbol.square(label_delta)
    
    output_sd <- mx.symbol.sqrt(mx.symbol.sum(output_delta))
    label_sd <- mx.symbol.sqrt(mx.symbol.sum(label_delta))
    
    numerator <- mx.symbol.sum(output_delta * label_delta)
    denominator <- output_sd * label_sd

    lro <- mx.symbol.MakeLoss(numerator/denominator)

    # Generate a new model
    model <- mx.model.FeedForward.create(symbol=lro, X=train.array, y=train.y, 
                                     num.round=5000, array.batch.size=1, optimizer = "adam",
                                     learning.rate = 0.0003, eval.metric = mx.metric.rmse,
                                     epoch.end.callback = mx.callback.log.train.metric(20, logger))
```
And I got this error:
```
Error in mx.model.init.params(symbol, input.shape, initializer, mx.cpu()) : 
Not enough information to get shapes
```
I tried to wrap the whole correlation formula in MXNet:
```
lro2 <- mx.symbol.MakeLoss(
    mx.symbol.negative((mx.symbol.sum(output * label) -
    (mx.symbol.sum(output) * mx.symbol.sum(label))) /
    mx.symbol.sqrt((mx.symbol.sum(mx.symbol.square(output)) -
    ((mx.symbol.sum(output)) * (mx.symbol.sum(output)))) *
    (mx.symbol.sum(mx.symbol.square(label)) - ((mx.symbol.sum(label)) * (mx.symbol.sum(label))))))
)
```
I can compile with this version, but my model runs very slowly and the code is apparently not very readable. I wonder if there is any way to implement get around the error and implement the first version as I described above.

**Comment by thirdwing** [comment]
I am afraid there is already error in your code to reshape the data:

```r
> train.x = matrix(data = rexp(200, rate = 10), nrow = 120, ncol = 6380)
> test.x = matrix(data = rexp(200, rate = 10), nrow = 60, ncol = 6380)
> train.y = matrix(data = rexp(200, rate = 10), nrow = 120, ncol = 319)
Warning message:
In matrix(data = rexp(200, rate = 10), nrow = 120, ncol = 319) :
  data length [200] is not a sub-multiple or multiple of the number of rows [120]
> test.y = matrix(data = rexp(200, rate = 10), nrow = 60, ncol = 319)
Warning message:
In matrix(data = rexp(200, rate = 10), nrow = 60, ncol = 319) :
  data length [200] is not a sub-multiple or multiple of the number of rows [60]
> train.array <-train.x
> dim(train.array) <-c(20,319,1,ncol(train.x))
Error in dim(train.array) <- c(20, 319, 1, ncol(train.x)) : 
  dims [product 40704400] do not match the length of object [765600]
```

Can you provide the code to reproduce `Not enough information to get shapes`?

**Comment by haimeh** [comment]
@thirdwing @nnguyen24 
 
If you replace the train.x and train.y array with:
```
# Generate testing data
train.x = matrix(data = rexp(200, rate = 10), nrow = 120, ncol = 6380)
train.y = matrix(data = rexp(200, rate = 10), nrow = 120, ncol = 320)

# Reshape testing data
train.array <- train.x
dim(train.array) <-c(20,6,1,ncol(X))

```
You can trigger the error.
The test set is not necessary to recreate the error.

I experienced the same error with my own fancy loss function.
I got around it, in the end, by implementing the loss inside a modified mx.model.train function (right before backward pass) replacing the symbols with mx.nd operations.

**Comment by thirdwing** [comment]
With the latest version of mxnet from github, the code works well:

```r
library(mxnet)

train.x = matrix(data = rexp(200, rate = 10), nrow = 120, ncol = 6380)
train.y = matrix(data = rexp(6380, rate = 10), nrow = 120, ncol = 319)

# Reshape testing data
train.array <- train.x
dim(train.array) <- c(319, 20, 120)
dim(train.y) <- c(319, 120)

data <- mx.symbol.Variable("data")

# Define the first fully connected layer
fc1 <- mx.symbol.FullyConnected(data, num_hidden = 100)
act.fun <- mx.symbol.Activation(fc1, act_type = "relu") # create a hidden layer with Rectified Linear Unit as its activation function.
output <- mx.symbol.FullyConnected(act.fun, num_hidden = 319)

# Customize loss function
label <- mx.symbol.Variable("label")

output_mean <- mx.symbol.mean(output)
label_mean <- mx.symbol.mean(label)

output_delta <- mx.symbol.broadcast_sub(output, output_mean)
label_delta <- mx.symbol.broadcast_sub(label, label_mean)

output_sqr <- mx.symbol.square(output_delta)
label_sqr <- mx.symbol.square(label_delta)

output_sd <- mx.symbol.sqrt(mx.symbol.sum(output_delta))
label_sd <- mx.symbol.sqrt(mx.symbol.sum(label_delta))

numerator <- mx.symbol.sum(output_delta * label_delta)
denominator <- output_sd * label_sd

lro <- mx.symbol.MakeLoss(numerator / denominator)

# Generate a new model
model <- mx.model.FeedForward.create(symbol = lro,
                                     X = train.array,
                                     y = train.y,
                                     num.round = 5000,
                                     array.batch.size = 1,
                                     optimizer = "adam",
                                     learning.rate = 0.0003,
                                     eval.metric = mx.metric.rmse,
                                     epoch.end.callback = mx.callback.log.train.metric(20))
```
```
Start training with 1 devices
[1] Train-rmse=NaN
.......
```

The output of `Makeloss` is the gradient, so the `mx.metric.rmse` produced NaN.

If you are using the prebuilt pkg, please wait for the update. I will update it soon.

**Comment by thirdwing** [comment]
The prebuilt pkg for Windows/OSX has been updated. This issue should have been fixed.

**Comment by nnguyen24** [comment]
Thank you @thirdwing , I will give it a try.

**Comment by haimeh** [comment]
Also, a mistake one might make, so for reference I am adding it here.
The shape inference may not always work with a chained composition such as:
  ```
discouragedSelfMatches <- mx.symbol.broadcast_mul(lhs = mx.symbol.broadcast_equal(lhs = distances,
                                                                                    rhs = mx.symbol.min(distances),
                                                    rhs = mx.symbol.max(distances))
```
but it will work if the symbols are separated, for instance:
  ```
selfMatches <- mx.symbol.broadcast_equal(lhs = distances,
                                           rhs = mx.symbol.min(distances))
  discouragedSelfMatches <- mx.symbol.broadcast_mul(lhs=selfMatches, 
                                                      rhs=mx.symbol.max(distances))
```

**Comment by thirdwing** [comment]
@haimeh Thank you for pointing out this.

I will look into it.
